app/services/hashtag_autocomplete_service.rb

```diff
@@ -117,8 +117,9 @@ def lookup(slugs, types_in_priority_order)
     # composer we want a slug without a suffix to be a category first, tag second.
     if slugs_without_suffixes.any?
       types_in_priority_order.each do |type|
-        found_from_slugs = set_refs(@@data_sources[type].lookup(guardian, slugs_without_suffixes))
-        found_from_slugs.each { |item| item.type = type }.sort_by! { |item| item.text.downcase }
+        found_from_slugs = lookup_for_type(type, guardian, slugs_without_suffixes)
+        next if !all_data_items_valid?(found_from_slugs)
+        found_from_slugs.sort_by! { |item| item.text.downcase }
         lookup_results[type.to_sym] = lookup_results[type.to_sym].concat(found_from_slugs)
```

Comment:
What do you think about extracting helper method to avoid duplications?

Something like
```
get_results!(lookup_results, type, guardian, slugs_without_suffixes)
get_results!(lookup_results, type, guardian, slugs_for_type)

def get_results!(results, type, guardian, slugs)
  found_items = lookup_for_type(type, guardian, slugs)
  bread if !all_data_items_valid?(found_items)
  found_items.sort_by! { |item| item.text.downcase }
  results[type.to_sym].concat(found_items)
end
```
concat should modify existing array, so I don't think we need assignment

Reply:
Yes I will try to make it more generic :)